packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+  CrowcountsTemp(CrowcountsTemp_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    CrowcountsTemp(i) = (Arowptrs(i + 1) - Arowptrs(i)) + (Browptrs(i + 1) - Browptrs(i));
+  }
+  const Values& Avals;
+  const RowPtrs& Arowptrs;
+  const ColInds& Acolinds;
+  const Values& Bvals;
+  const RowPtrs& Browptrs;
+  const ColInds& Bcolinds;
+  RowPtrs& CrowcountsTemp;
+};
+
+template<typename Scalar, typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct UnmergedSumFunctor
+{
+  UnmergedSumFunctor(const Values& Avals_, const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const Scalar& scalarA_,
+           const Values& Bvals_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, const Scalar& scalarB_,
+           Values& CvalsOver_, const RowPtrs& CrowptrsTemp_, ColInds& CcolindsOver_) :
+    Avals(Avals_), Arowptrs(Arowptrs_), Acolinds(Acolinds_), scalarA(scalarA_),
+    Bvals(Bvals_), Browptrs(Browptrs_), Bcolinds(Bcolinds_), scalarB(scalarB_),
+    CvalsOver(CvalsOver_), CrowptrsTemp(CrowptrsTemp_), CcolindsOver(CcolindsOver_)
+  {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    Ordinal inserted = 0;
+    auto rowStart = CrowptrsTemp(i);
+    auto arowstart = Arowptrs(i);
+    auto arowlen = Arowptrs(i + 1) - arowstart;
+    auto browstart = Browptrs(i);
+    auto browlen = Browptrs(i + 1) - browstart;
+    for(Ordinal j = 0; j < arowlen; j++)
+    {
+      CvalsOver(rowStart + inserted) = scalarA * Avals(arowstart + j);
+      CcolindsOver(rowStart + inserted) = Acolinds(arowstart + j);
+      inserted++;
+    }
+    for(Ordinal j = 0; j < browlen; j++)
+    {
+      CvalsOver(rowStart + inserted) = scalarB * Bvals(browstart + j);
+      CcolindsOver(rowStart + inserted) = Bcolinds(browstart + j);
+      inserted++;
+    }
+  }
+  const Values& Avals;
+  const RowPtrs& Arowptrs;
+  const ColInds& Acolinds;
+  const Scalar scalarA;
+  const Values& Bvals;
+  const RowPtrs& Browptrs;
+  const ColInds& Bcolinds;
+  const Scalar scalarB;
+  Values& CvalsOver;
+  const RowPtrs& CrowptrsTemp;
+  ColInds& CcolindsOver;
+};
+
+template<typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct SortAndMergeFunctor
+{
+  SortAndMergeFunctor(const RowPtrs& CrowptrTemp_, RowPtrs& Crowcounts_, ColInds& CcolindOver_, Values& CvalsOver_, ColInds& CcolindAux_, Values& CvalsAux_, size_t numCols_) :
+    CrowptrTemp(CrowptrTemp_),
+    Crowcounts(Crowcounts_),
+    CcolindOver(CcolindOver_),
+    CvalsOver(CvalsOver_),
+    CcolindAux(CcolindAux_),
+    CvalsAux(CvalsAux_)
+  {
+    //Compute the minimum number of radix sort iterations needed to sort values
+    //in the range [0, numCols)
+    //If unsorted kernel called directly, Ordinal == LO and numCols is num local cols
+    numRadixSweeps = 0;
+    auto temp = numCols_;
+    while(temp)
+    {
+      numRadixSweeps++;
+      temp >>= 4;
+    }
+  }
+  //radixSort2: a CUDA-friendly equivalent for Tpetra::sort2
```

Comment:
@mhoemmen I used the shellSort and it slowed down the sort+merge functor by about 65%, so I will move the radix sort into that file and document/test it.